lib/irb.rb

```diff
@@ -21,36 +21,917 @@
+#    and by direct assignments in the configuration file.
+# 1. Assigns the context to variable +conf+.
+# 1. Assigns command-line arguments to variable <tt>ARGV</tt>.
+# 1. Prints the {prompt}[rdoc-ref:IRB@Prompt+and+Return+Formats].
+# 1. Puts the content of the
+#    {initialization script}[rdoc-ref:IRB@Initialization+Script]
+#    onto the \IRB shell, just as if it were user-typed commands.
+#
+# === The Command Line
+#
+# On the command line, all options precede all arguments;
+# the first item that is not recognized as an option is treated as an argument,
+# as are all items that follow.
+#
+# ==== Command-Line Options
+#
+# Many command-line options affect entries in hash <tt>IRB.conf</tt>,
+# which in turn affect the initial configuration of the \IRB session.
+#
+# Details of the options are described in the relevant subsections below.
+#
+# A cursory list of the \IRB command-line options
+# may be seen in the {help message}[rdoc-ref:lib/irb/lc/help-message],
+# which is also displayed if you use command-line option <tt>--help</tt>.
+#
+# If you are interested in a specific option, consult the
+# {index}[./doc/irb/indexes_rdoc.html#label-Index+of+Command-Line+Options].
+#
+# ==== Command-Line Arguments
+#
+# Command-line arguments are passed to \IRB in array +ARGV+:
+#
+#   $ irb --noscript Foo Bar Baz
+#   irb(main):001> ARGV
+#   => ["Foo", "Bar", "Baz"]
+#
+# Command-line option <tt>--</tt> causes everything that follows
+# to be treated as arguments, even those that look like options:
+#
+#   $ irb --noscript -- --noscript -- Foo Bar Baz
+#   irb(main):001> ARGV
+#   => ["--noscript", "--", "Foo", "Bar", "Baz"]
+#
+# === Configuration File
+#
+# You can initialize \IRB via a <i>configuration file</i>.
+#
+# If command-line option <tt>-f</tt> is given,
+# no configuration file is looked for.
+#
+# Otherwise, \IRB reads and interprets a configuration file
+# if one is available.
+#
+# The configuration file can contain any Ruby code, and can usefully include
+# user code that:
+#
+# - Can then be debugged in \IRB.
+# - Configures \IRB itself.
+# - Requires or loads files.
+#
+# The path to the configuration file is the first found among:
+#
+# - The value of variable <tt>$IRBRC</tt>, if defined.
+# - The value of variable <tt>$XDG_CONFIG_HOME/irb/irbrc</tt>, if defined.
+# - File +.config/irb/irbrc+, if it exists.
+# - File +.irbrc+, if it exists.
+# - File +irb.rc+, if it exists.
+# - File +_irbrc+, if it exists.
+# - The value of variable <tt>$irbrc</tt>, if defined.
+#
+# If the search fails, there is no configuration file.
+#
+# If the search succeeds, the configuration file is read as Ruby code,
+# and so can contain any Ruby programming you like.
+#
+# \Method <tt>conf.rc?</tt> returns +true+ if a configuration file was read,
+# +false+ otherwise.
+# \Hash entry <tt>IRB.conf[:RC]</tt> also contains that value.
+#
+# [REVIEWERS] What should be said about IRB.conf[:RC_NAME_GENERATOR]?
+#
+# === Session \Context
```

Comment:
I think we should move most of the documentation after this related to `IRB::Context` into the documentation for `IRB::Context`.

Reply:
I'm not clear on what exactly you think should be moved.  As I said, I'm trying to document as features, rather than implementation (options, IRB.conf entries -- and also classes).  For each feature I've tried to show how it's state is initialized, and what if anything can modify it downstream.  I'd like to leave all here, at least for now.

Reply:
I think we should remove the `Context` section as I consider it a private component and I don't want to maintain backward-compatibility for it.

Reply:
Removed.